### ISSUE TITLE
[feature idea] unit tests

### ISSUE THREAD

**Issue by Almenon** [microsoft/vscode-extension-samples#91]
It would be nice if some of the samples had unit tests or there was a sample unit test project. It's good for people looking through the samples to see how to unit test it because putting off unit tests just leads to tech debt in the future.

**Comment by lannonbr** [comment]
I would be interested in making a PR for this given I've been writing a vscode extension and setting up some unit tests for it. Not sure what to test for an example but I am interested in helping out

**Comment by gkalpak** [comment]
It would also be useful to distinguish between unit tests and end-to-end/integration tests.

E2E tests are useful because they runs on an actual VSCode instance, but they are too slow to be run often during development (e.g. on save).

The unit tests are very fast and suitable for immediate feedback during development (e.g. on save), but require mocking; especially the `vscode` parts, which is tricky until you figure it out the first time.

##
FWIW, I am using [mock-require][1] for my setup. Here is my [`vscode` mock][2] and here my [overall test setup][3] (which is a little involved, because I wanted to use jasmine :grin:).

[1]: https://www.npmjs.com/package/mock-require
[2]: https://github.com/gkalpak/aio-docs-utils/blob/4f0d1a760ac3882b0e553fdaef8e7405eedfdf65/src/test/helpers/vscode.mock.ts
[3]: https://github.com/gkalpak/aio-docs-utils/tree/4f0d1a760ac3882b0e553fdaef8e7405eedfdf65/src/test

**Comment by Almenon** [comment]
I just updated my unit tests for the new webview API, the same format could be applied to the webview sample.

https://github.com/Almenon/AREPL-vscode/blob/master/test/extension.test.ts

The only catch is I had to make a mock context like so:
```typescript
    const arepl = vscode.extensions.getExtension("almenon.arepl")!;

    const mockContext: any = {
        asAbsolutePath: (file: string)=>{
            return __dirname + "/" + file
        },
        extensionPath: arepl.extensionPath,
    }
```

Aside from that it's just calling the function that updates the html and verifying that the expected output is in the html, pretty simple.

**Comment by Almenon** [comment]
A person just asked about this in the vscode slack channel.  

```
Anon1
guys what do you do about unit testing `vscode` ?
We’re currently writing stub methods for everything we use
that’s getting boring, real quick
```

Some responses:

> Anon2 
doing the same over here. it’s not too bad but maybe i’m not using as much of the API as you are. i’m also using sinon for the stubs.

> Anon3 
It depends on what code you're trying to write tests for. You're right in that if you want to do unit testing you'll end up writing stubs for vscode api's.
In my experience it's a case by case evaluation on which approach (unit vs integration) will provide you the most benefits.


> Anon4 
I had noticed the same thing. Most extensions do not have tests. In my re-write of my extension I am making it a point to have unit tests. I’m using jest and have mocked the vscode api. I only mock what I need when I need it so that I don’t get ahead of myself. https://github.com/kenhowardpdx/vscode-gist/

In summary it seems like most people are in favor of mocking the actual VSCode API. This lets your tests run faster at the cost of the time it takes to make the mock.

**Comment by connor4312** [comment]
Duplicates https://github.com/microsoft/vscode/issues/94746

**Comment by Almenon** [comment]
This issue is a subset of microsoft/vscode#94746, but not a duplicate. microsoft/vscode#94746 is asking for a mock library and general guidance. Can you reopen please?

**Comment by connor4312** [comment]
There is guidance given to mocking in the discussion on that issue, for example my comment here: https://github.com/microsoft/vscode/issues/94746#issuecomment-615297619

**Comment by Almenon** [comment]
I appreciate the comment, but it's not the same as having multiple unit test examples to reference.
